src/gui/painting/qpainter.cpp

```diff
@@ -2673,6 +2673,182 @@ QPainterPath QPainter::clipPath() const
+
+    \sa setClipPath(), clipRegion(), setClipping()
+*/
+QPainterPath QPainter::clipPathF() const
+{
+    Q_D(const QPainter);
+
+    if (!d->engine) {
+        qWarning("QPainter::clipPath: Painter not active");
+        return QPainterPath();
+    }
+
+    // No clip, return empty
+    if (d->state->clipInfo.isEmpty()) {
+        return QPainterPath();
+    } else {
+
+        // Update inverse matrix, used below.
+        if (!d->txinv)
+            const_cast<QPainter *>(this)->d_ptr->updateInvMatrix();
+
+        // For the simple case avoid conversion.
+        if (d->state->clipInfo.size() == 1
+            && d->state->clipInfo.at(0).clipType == QPainterClipInfo::PathClip) {
+            QTransform matrix = (d->state->clipInfo.at(0).matrix * d->invMatrix);
+            return d->state->clipInfo.at(0).path * matrix;
+
+        } else if (d->state->clipInfo.size() == 1
+                   && d->state->clipInfo.at(0).clipType == QPainterClipInfo::RectClip) {
+            QTransform matrix = (d->state->clipInfo.at(0).matrix * d->invMatrix);
+            QPainterPath path;
+            path.addRect(d->state->clipInfo.at(0).rect);
+            return path * matrix;
+        } else {
+            /*
+            * Modified version of "clipRegion" that is used in ::clipPath
+            * as "return qt_regionToPath(clipRegion());" that prevents rounding of values,
+            * placing them directly into a QPainterPath.
+            */
+            QPainterPath path;
+            bool lastWasNothing = true;
+            if (!d->txinv)
+                const_cast<QPainter *>(this)->d_ptr->updateInvMatrix();
+            // ### Falcon: Use QPainterPath
+            for (const QPainterClipInfo &info : qAsConst(d->state->clipInfo)) {
+                switch (info.clipType) {
+
+                    case QPainterClipInfo::PathClip: {
+                        QTransform matrix = (info.matrix * d->invMatrix);
+                        if (lastWasNothing) {
+                            path = info.path * matrix;
+                            lastWasNothing = false;
+                            continue;
+                        }
+                        if (info.operation == Qt::IntersectClip) {
+                            path &= (info.path * matrix);
+                        } else if (info.operation == Qt::NoClip) {
+                            lastWasNothing = true;
+                            path = QPainterPath();
+                        } else {
+                            path = info.path * matrix;
+                        }
+                        break;
+                    }
+                                                 
+                    case QPainterClipInfo::RectClip: {
+                        QTransform matrix = (info.matrix * d->invMatrix);
+                        if (lastWasNothing) {
+                            QPainterPath tempPath;
+                            tempPath.addRect(info.rect);
```

Comment:
`addRect` only accepts `QRectF` so you'll need to create a `QRectF` from this rect.

Reply:
I'm confused about how line 2718 above works as that seems contrary to what I think should work.